doc/design/resource-dependency.md

```diff
@@ -0,0 +1,343 @@
+      FieldPath string `json:"fieldPath"`
+    }
+
+    type ResourceReference struct {
+      APIVersion string `json:"apiVersion"`
+
+      Kind string `json:"kind"`
+
+      // Name indicates the name of the object in the same namespace.
+      Name string `json:"name,omitempty"`
+    }
+
+    type DataOutputStatus struct {
+      // Ready indicates whether the field data of the source resource is ready.
+      Ready bool `json:"ready"`
+    }
+    ```
+
+    For example, a mysql connection secret DataOutput coming from RDSInstance my-rds would look like:
+
+    ```yaml
+    apiVersion: experimental.core.oam.dev/v1alpha2
+    kind: DataOutput
+    metadata:
+      name: mysql-conn-secret
+    spec:
+      resource:
+        apiVersion: alibaba.oam.crossplane.io/v1alpha1
+        kind: RDSInstance
+        name: my-rds
+      fieldPath: "status.connSecret"
+    ```
+
+    In this case, the OAM runtime will automatically build a dependency between the `my-web` and `my-rds`.
+
+    It will wait until `my-rds` object's field `status.connSecret` has value, and marks the DataOutput status ready:
+
+    ```yaml
+    apiVersion: alibaba.oam.crossplane.io/v1alpha1
+    kind: RDSInstance
+    metadata:
+      name: my-rds
+    spec:
+      ...
+    status:
+      connSecret: mysql-conn # OAM runtime will wait for this field to have value
+    ---
+    apiVersion: experimental.core.oam.dev/v1alpha2
+    kind: DataOutput
+    metadata:
+      name: mysql-conn-secret
+    ...
+    status:
+      ready: true # once the above connSecret has value this will be marked true
+    ```
+
+    At this point, the dependency has been satisfied.
+    The OAM runtime will create `my-web` object with corresponding values passed to its fields.
+    
+    ```yaml
+    kind: MyWorkload
+    metadata:
+      name: my-web
+      ..
+    spec:
+      connSecret: mysql-conn # patched by OAM runtime
+    ```
+
+3. Below we show how to combine the above concepts (DataInput, DataOutput) into OAM.
+
+    In OAM Component we put a syntax-sugar on existing `Parameters` fields:
+    as long as the name of the Parameter matches the name of the DataOutput, it becomes a DataInput.
+    For example, the following parameter mysql-conn-secret in Component matches the name of DataOutput trait.
+
+    ```yaml
+    kind: Component
+    metadata:
+      name: my-web
+    spec:
+      workload:
+        kind: MyWorkload
+        metadata:
+          name: my-web
+      parameters:
+      - name: mysql-conn-secret # this matches the name of DataOutput
+        fieldPaths: ["spec.connSecret"]
+    ---
+    kind: Component
+    metadata:
+      name: my-rds
+    spec:
+      workload:
+        kind: RDSInstance
+        metadata:
+          name: my-rds
+    ---
+    # In ApplicationConfiguration
+    components:
+    - componentName: my-web
+      traits:
+      - trait:
+          apiVersion: experimental.core.oam.dev/v1alpha2
+          kind: DataOutput
```

Comment:
A component binding with a `DataOutput` trait seems tell others this component will output something. But in fact the component is consuming data from this trait. So I think naming is with `DataInput` is more intuitive.

Reply:
same as above. Gonna resolve this and merge conversation.